### PR TITLE
Turn channelid from bytes to uint256

### LINKED ISSUE #1940: Adjust proxies for channel identifier changes done for raiden-contracts 193
After the solution for [raiden-contracts 193](https://github.com/raiden-network/raiden-contracts/issues/193) gets implemented we will need to adjust the raiden smart contract proxies in the raiden client in order to be able to call the `TokenNetwork` contract functions with the new interface changes.

For details on how the contracts will change look at the [issue description](https://github.com/raiden-network/raiden-contracts/issues/193#issue-343704868) and the eventual PR [TODO: Add link when opened].

**Important note**

The change to the proxies should be only as to how the proxies make the actual calls to the on-chain contract. The rest of the interface with the raiden client should remain the same. Essentially we should add the `channel_identifier` to all `TokenNetwork` calls.

### LINKED ISSUE #2000: Adjust the codebase for the channel identifier change to uint256
## Problem Definition

After all the changes proposed by https://github.com/raiden-network/raiden-contracts/issues/193
we will need to adjust the codebase to use the new `uint256` `channel_identifier`.

## Task

Adjust the codebase for the new `channel_identifier`.
This needs some thinking so the following list may be non-exchaustive:

1. [ ] Change the type of the channel_identifier in the [typing](https://github.com/raiden-network/raiden/blob/master/raiden/utils/typing.py#L26) module.
2. [ ] Grep for the above type and follow the codepaths to make sure the appropriate type is used
3. [ ] Adjust the channel_identifier in [messages.py.](https://github.com/raiden-network/raiden/blob/4e0bfe55daea6f65ce439bc7a359b629dc0ae7cb/raiden/messages.py#L76) and make sure that a uint256 is used for signing
4. [ ] Adjust the tests where needed

### PR DESCRIPTION
**Pull request by LefterisJP**
Fix #2000 and fix #1940

Incorporates the initial workd @err508 did for #1940 at https://github.com/raiden-network/raiden/pull/2018

It takes it from there and modifies the entire codebase to make sure channel_identifier is uint256 instead of bytes.

Also adapts the proxies and generally the interface to the smart contracts to make sure they are compatible with the new smart contract changes introduced for the [Red Eyes Testnet 4](https://github.com/raiden-network/raiden-contracts/milestone/5) milestone.

### CODE REVIEW

File: raiden/network/proxies/token_network.py [hunk 1]
```diff
@@ -52,6 +52,17 @@
 
 log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
 
+ParticipantDetails = Dict[
```
Comment (andrevmatos) [inline raiden/network/proxies/token_network.py]: Shouldn't this kind of type-alias be in `raiden.utils.typing`?

Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: no. not all type aliases need to be in typing if they are local to the module they should stay in the module.

Also note: This PR is **NOT** ready for review. I am pushing only when I wanna move stuff from desktop to laptop or vice versa and when I want to run travis.

**Comment by LefterisJP** [comment]
All integration tests pass, except for the freaking [echo node test](https://travis-ci.org/raiden-network/raiden/jobs/414004410).

This PR is ready for review finally :)

File: raiden/api/v1/encoding.py [hunk 2]
```diff
@@ -242,7 +242,7 @@ class Meta:
 
 
 class ChannelStateSchema(BaseSchema):
-    channel_identifier = KeccakField(attribute='identifier')
```
Comment (palango) [inline raiden/api/v1/encoding.py]: `KeccakField` is unused afterwards and can be removed

File: raiden/network/proxies/token_network.py [hunk 3]
```diff
@@ -194,26 +205,65 @@ def _new_netting_channel(self, partner: typing.Address, settle_timeout: int):
...
+                'getChannelIdentifier',
+                to_checksum_address(participant1),
+                to_checksum_address(participant2),
+            )
+        assert isinstance(channel_identifier, typing.T_ChannelID)
+        if channel_identifier == 0:
+            raise ChannelIncorrectStateError(
+                f'When calling {called_by_fn} either 0 value was given for the '
+                'channel_identifier or getChannelIdentifier returned 0, meaning '
+                f'no channel is currently open between {pex(participant1)} and '
```
Comment (palango) [inline raiden/network/proxies/token_network.py]: 'open'  could be misleading here, maybe use exists?

File: raiden/network/proxies/token_network.py [hunk 4]
```diff
@@ -52,6 +52,17 @@
 
 log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
 
+ParticipantDetails = Dict[
+    str,
+    typing.Union[
```
Comment (hackaugusto) [inline raiden/network/proxies/token_network.py]: ~~This is the wrong type constructor, [Union](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#union-types) builds a sum and not product type, we want to use [Tuple](https://mypy.readthedocs.io/en/stable/builtin_types.html#built-in-types) here~~

This is a bad type IMO, we should just use a namedtuple or a dataclass, which have a limited number of fields and we can individually type them.

File: raiden/network/proxies/token_network.py [hunk 5]
```diff
@@ -223,27 +273,44 @@ def detail_participant(
             'nonce': data[4],
         }
 
-    def detail_channel(self, participant1: typing.Address, participant2: typing.Address) -> Dict:
-        """ Returns a dictionary with the channel specific information. """
+    def detail_channel(
+            self,
+            participant1: typing.Address,
+            participant2: typing.Address,
+            channel_identifier: typing.Optional[typing.ChannelID] = None,
```
Comment (palango) [inline raiden/network/proxies/token_network.py]: When setting the default to `None` you don't need to write `Optional[XXX]`.

File: raiden/network/proxies/token_network.py [hunk 6]
```diff
@@ -194,26 +205,65 @@ def _new_netting_channel(self, partner: typing.Address, settle_timeout: int):
...
+                f'no channel is currently open between {pex(participant1)} and '
+                f'{pex(participant2)}',
+            )
+        return channel_identifier
+
+    def channel_exists_and_not_settled(
+            self,
+            participant1: typing.Address,
+            participant2: typing.Address,
+            channel_identifier: typing.Optional[typing.ChannelID] = None,
```
Comment (hackaugusto) [inline raiden/network/proxies/token_network.py]: some functions have the channel_identifier as the first argument, while other as the last, can we be consistent with this?

File: raiden/network/proxies/token_network.py [hunk 7]
```diff
@@ -223,27 +273,44 @@ def detail_participant(
...
 
-        assert isinstance(channel_data[0], typing.T_ChannelID)
-
         return {
-            'channel_identifier': channel_data[0],
-            'settle_block_number': channel_data[1],
-            'state': channel_data[2],
+            'channel_identifier': channel_identifier,
+            'settle_block_number': channel_data[0],
+            'state': channel_data[1],
```
Comment (hackaugusto) [inline raiden/network/proxies/token_network.py]: what about giving a name for these constants? preferably in the raiden-contracts repo

File: raiden/network/blockchain_service.py [hunk 8]
```diff
@@ -174,8 +174,8 @@ def payment_channel(
 
         if not is_binary_address(token_network_address):
             raise ValueError('address must be a valid address')
-        if not ishash(channel_id):
-            raise ValueError('identifier must be a hash')
+        if not isinstance(channel_id, T_ChannelID):
+            raise ValueError('channel identifier must be a hash')
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/network/blockchain_service.py]: I guess so ;). Will fix.

File: raiden/connection_manager.py [hunk 9]
```diff
@@ -286,7 +286,7 @@ def _join_partner(self, partner: Address):
             log.exception('connection manager: deposit failed')
         except ChannelIncorrectStateError:
             log.exception('connection manager: channel not in opened state')
-        except InsufficientFunds as e:
+        except InsufficientFunds:
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/connection_manager.py]: Sure, but this is not a problem in my PR. I just removed an unused variable. Flake8 seems to complain for more stuff locally than it does in travis. I can add it back and print the exception info if needed. Will check what that would be first though.

File: raiden/tests/integration/contracts/test_token_network.py [hunk 10]
```diff
@@ -166,60 +177,72 @@ def test_token_network_proxy_basics(
...
-        0,
-        EMPTY_HASH,
+        channel_identifier=channel_identifier,
+        transferred_amount=0,
+        locked_amount=0,
+        locksroot=EMPTY_HASH,
+        partner=c1_client.sender,
+        partner_transferred_amount=transferred_amount,
+        partner_locked_amount=0,
+        partner_locksroot=EMPTY_HASH,
```
Comment (palango) [inline raiden/tests/integration/contracts/test_token_network.py]: 👍 for adding the keywords for the arguments.

File: raiden/network/proxies/token_network.py [hunk 11]
```diff
@@ -531,8 +652,8 @@ def close(
             self,
             channel_identifier: typing.ChannelID,
             partner: typing.Address,
-            nonce: typing.Nonce,
             balance_hash: typing.BalanceHash,
+            nonce: typing.Nonce,
```
Comment (hackaugusto) [inline raiden/network/proxies/token_network.py]: why did you change the order of the arguments? this can lead to bugs if not all places are updated.

File: raiden/transfer/mediated_transfer/events.py [hunk 12]
```diff
@@ -4,32 +4,48 @@
...
+from raiden.transfer.state import BalanceProofUnsignedState, HashTimeLockState
 from raiden.utils import pex, sha3
+from raiden.utils import typing
+
+
+# According to the smart contracts as of 07/08:
+# https://github.com/raiden-network/raiden-contracts/blob/fff8646ebcf2c812f40891c2825e12ed03cc7628/raiden_contracts/contracts/TokenNetwork.sol#L213
+# channel_identifier can never be 0. We make this a requirement in the client and use this fact
+# to signify that a channel_identifier of `0` passed to the messages adds them to the
+# global queu
```
Comment (palango) [inline raiden/transfer/mediated_transfer/events.py]: /s/queu/queue

File: raiden/network/proxies/token_network.py (hunk 5 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: I know I know, but I find it a lot more readable when the code is explicit

Comment (palango) [reply raiden/network/proxies/token_network.py]: Fine for me!

File: raiden/network/proxies/token_network.py (hunk 11 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: I know ... It's something @err508 started in his first PR and I also commented against doing [here](https://github.com/raiden-network/raiden/pull/2018#discussion_r2070552450).

His reasoning was that he wanted same order as the contract arguments. The code he had introduced had bugs as it had not changed all occurences. I just rolled with it in the end but I made sure that every single occurrence is changed and is also called with kwargs so the order no longer matter.

File: raiden/network/proxies/token_network.py (hunk 4 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: > This is a bad type IMO

This is the type that the function was returning before my PR.

I can change it and make it a named tuple, but that's just additional changes not really related to the PR.

File: raiden/network/proxies/token_network.py (hunk 6 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: The `channel_identifier` is consistently last for all the functions where it's optional. If optional the latest channel identifier is requested by asking the proxy for the current channel identifier between `participant1` and `participant2`

File: raiden/network/proxies/token_network.py (hunk 7 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: Sure

File: raiden/network/proxies/token_network.py (hunk 6 above)
Comment (hackaugusto) [reply raiden/network/proxies/token_network.py]: > The channel_identifier is consistently last for all the functions where it's optional

https://github.com/raiden-network/raiden/pull/2035/files#diff-d02574a7a7eca826412780e80a6326a3R210

Comment (hackaugusto) [reply raiden/network/proxies/token_network.py]: also, if the default value is `None` mypy will infer that the type is optional, so this signature is suficient:

`channel_identifier: typing.ChannelID = None`

File: raiden/api/v1/encoding.py (hunk 2 above)
Comment (LefterisJP) [reply raiden/api/v1/encoding.py]: removed it

**Comment by rakanalh** [comment]
reviewing now

File: raiden/network/proxies/token_network.py [hunk 13]
```diff
@@ -194,56 +219,113 @@ def _new_netting_channel(self, partner: typing.Address, settle_timeout: int):
...
+            nonce=data[ParticipantInfoIndex.NONCE],
+        )
+
+    def detail_channel(
+            self,
+            participant1: typing.Address,
+            participant2: typing.Address,
+            channel_identifier: typing.ChannelID = None,
+    ) -> ChannelData:
+        """ Returns a dictionary with the channel specific information.
```
Comment (rakanalh) [inline raiden/network/proxies/token_network.py]: Returns an instance of `ChannelData` could be a more accurate description

File: raiden/network/proxies/token_network.py [hunk 14]
```diff
@@ -194,56 +219,113 @@ def _new_netting_channel(self, partner: typing.Address, settle_timeout: int):
...
             self,
-            participant1: typing.Address,
-            participant2: typing.Address,
-    ) -> Dict:
+            channel_identifier: typing.ChannelID,
+            participant: typing.Address,
+            partner: typing.Address,
+    ) -> ParticipantDetails:
         """ Returns a dictionary with the channel participant information. """
+
```
Comment (rakanalh) [inline raiden/network/proxies/token_network.py]: returns `ParticipantDetails`

File: raiden/network/proxies/token_network.py [hunk 15]
```diff
@@ -194,56 +219,113 @@ def _new_netting_channel(self, partner: typing.Address, settle_timeout: int):
...
+            state=channel_data[ChannelInfoIndex.STATE],
+        )
 
     def detail_participants(
             self,
             participant1: typing.Address,
             participant2: typing.Address,
-    ) -> Dict:
+            channel_identifier: typing.ChannelID = None,
+    ) -> ParticipantsDetails:
```
Comment (rakanalh) [inline raiden/network/proxies/token_network.py]: Returns `ParticipantDetails` instead of "dictionary"

File: raiden/network/proxies/token_network.py [hunk 16]
```diff
@@ -255,24 +337,23 @@ def detail_participants(
...
+        partner_data = self.detail_participant(channel_identifier, participant2, participant1)
+        return ParticipantsDetails(our_details=our_data, partner_details=partner_data)
 
-    def detail(self, participant1: typing.Address, participant2: typing.Address) -> Dict:
+    def detail(
+            self,
+            participant1: typing.Address,
+            participant2: typing.Address,
+            channel_identifier: typing.ChannelID = None,
+    ) -> ChannelDetails:
```
Comment (rakanalh) [inline raiden/network/proxies/token_network.py]: Returns `ChannelDetails`

File: raiden/network/proxies/token_network.py [hunk 17]
```diff
@@ -302,101 +387,130 @@ def settlement_timeout_max(self) -> int:
...
     def settle_block_number(
             self,
             participant1: typing.Address,
             participant2: typing.Address,
-    ) -> typing.BlockNumber:
-        """ Returns the channel settle_block_number. """
-        channel_data = self.detail_channel(participant1, participant2)
-        return channel_data.get('settle_block_number')
+            channel_identifier: typing.ChannelID = None,
+    ) -> typing.Optional[typing.BlockNumber]:
```
Comment (rakanalh) [inline raiden/network/proxies/token_network.py]: General question:
in `CONTRIBUTING.md` it was mentioned that if the object takes `None` as a default value then we don't type it as `Optional`. Does this mean we only type return values which could be `None` as `Optional`?

File: requirements-lint.txt [hunk 18]
```diff
@@ -1,3 +1,9 @@
+# pycodestyle==2.4.0
+# To avoid [this](https://github.com/PyCQA/pycodestyle/issues/635) bug
+# we need pycodestyle 2.4.0 but unfortunately this does not work with flake8
+# due to https://github.com/PyCQA/pycodestyle/issues/741
+# Once a new version of flake8 is released we can specify the latest pycodestyle
+# version here
```
Comment (rakanalh) [inline requirements-lint.txt]: Should we add a follow-up issue for this?

Comment (LefterisJP) [reply requirements-lint.txt]: Naah the minute flake8 is released this is solved and we can update it and remove this comment

File: raiden/network/proxies/token_network.py (hunk 17 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: This is for them being as arguments. Here it's a return type. Can't put `= None` here

File: raiden/network/proxies/token_network.py (hunk 16 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: ?

File: raiden/network/proxies/token_network.py (hunk 15 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: ??

File: raiden/network/proxies/token_network.py (hunk 13 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: Aha ... I see .. now .. your comment was on the wrong line. You mean to edit the docstrings. Will do.

File: raiden/network/proxies/token_network.py (hunk 17 above)
Comment (rakanalh) [reply raiden/network/proxies/token_network.py]: I think you misunderstood the question. So this function might return None... but we choose  to use optional here while we don't in func arguments where the passed value or default value could be None. Don't you see this as an inconsistency? We might want to define the use of Optional as a type for return values but not parameters in this case.

File: raiden/network/proxies/token_network.py (hunk 13 above)
Comment (rakanalh) [reply raiden/network/proxies/token_network.py]: Yea, Github doesn't let you add comments on un-modified lines.

File: raiden/network/proxies/token_network.py (hunk 17 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: No I don't see this as an incosistency.

**Comment by LefterisJP** [comment]
I am merging this. Not waiting till all tests finish due to the changes being too minimal to affect tests.
